### PR TITLE
KAFKA-16448: Unify error-callback exception handling

### PR DESCRIPTION
**Pull request by mjsax**
@sebastienviale @loicgreffier -- as discussed on other PR review, I think we should apply some cleanup and unification for the different handler. Would love to get your feedback.

\cc @cadonna @lucasbru @bbejeck

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java [hunk 1]
```diff
@@ -81,6 +81,17 @@ public TaskId taskId() {
...
+    @Override
+    public String toString() {
+        return "ErrorHandlerContext{" +
+            "topic='" + topic + '\'' +
+            ", partition=" + partition +
+            ", offset=" + offset +
+            ", processorNodeId='" + processorNodeId + '\'' +
+            ", taskId=" + taskId +
+            '}';
+    }
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java:94]: Omitting `headers` on purpose, as we should not log user-data.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java [hunk 2]
```diff
@@ -330,7 +330,8 @@ private void reprocessState(final List<TopicPartition> topicPartitions,
                                 Thread.currentThread().getName(),
                                 globalProcessorContext.taskId().toString(),
                                 globalProcessorContext.metrics()
-                            )
+                            ),
+                            null
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:334]: Side cleanup as discussed on some other PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java [hunk 3]
```diff
@@ -202,7 +202,7 @@ public void process(final Record<KIn, VIn> record) {
         } catch (final FailedProcessingException | TaskCorruptedException | TaskMigratedException e) {
             // Rethrow exceptions that should not be handled here
             throw e;
-        } catch (final RuntimeException e) {
+        } catch (final RuntimeException processingException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java:206]: side cleanup: use proper variable names

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 4]
```diff
@@ -205,7 +207,7 @@ public <K, V> void send(final String topic,
                 key,
                 keySerializer,
                 exception);
-        } catch (final Exception exception) {
+        } catch (final RuntimeException serializationException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:213]: side cleanup: use proper variable names

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 5]
```diff
@@ -372,24 +380,24 @@ private <V> StreamsException createStreamsExceptionForValueClassCastException(fi
     }
 
     private void recordSendError(final String topic,
-                                 final Exception exception,
+                                 final Exception productionException,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:378]: side cleanup: use proper variable names

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 6]
```diff
@@ -406,17 +414,31 @@ private void recordSendError(final String topic,
                     taskId
                 );
 
-                if (productionExceptionHandler.handle(errorHandlerContext, serializedRecord, exception) == ProductionExceptionHandlerResponse.FAIL) {
+                final ProductionExceptionHandlerResponse response;
+                try {
+                    response = productionExceptionHandler.handle(errorHandlerContext, serializedRecord, productionException);
+                } catch (final RuntimeException fatalUserException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:418]: new: handle exception from `ProductionExceptionHandler` callback

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordDeserializer.java [hunk 7]
```diff
@@ -69,47 +70,42 @@ ConsumerRecord<Object, Object> deserialize(final ProcessorContext<?, ?> processo
...
         }
     }
 
     public static void handleDeserializationFailure(final DeserializationExceptionHandler deserializationExceptionHandler,
                                                     final ProcessorContext<?, ?> processorContext,
-                                                    final Exception deserializationException,
-                                                    final ConsumerRecord<byte[], byte[]> rawRecord,
-                                                    final Logger log,
-                                                    final Sensor droppedRecordsSensor) {
-        handleDeserializationFailure(deserializationExceptionHandler, processorContext, deserializationException, rawRecord, log, droppedRecordsSensor, null);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordDeserializer.java:84]: Side cleanup as discussed on some other PR

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 8]
```diff
@@ -915,6 +917,13 @@ public void punctuate(final ProcessorNode<?, ?, ?, ?> node,
 
         try {
             maybeMeasureLatency(() -> punctuator.punctuate(timestamp), time, punctuateLatencySensor);
+        } catch (final TimeoutException timeoutException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:930]: Totally unrelated (might actually to good to put into different PR, but want to avoid that I forget it).

Seems we forgot to handle this case entirely... we might call `context.forward()` inside of  `punctuate()` and thus need to handle the same error cases as for `node.process()`...

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 9]
```diff
@@ -293,27 +295,33 @@ private <K, V> void handleException(final ProductionExceptionHandler.Serializati
...
+                    errorHandlerContext
+                ),
+                serializationException
             );
-            response = productionExceptionHandler.handleSerializationException(errorHandlerContext, record, exception, origin);
-        } catch (final Exception e) {
-            log.error("Fatal when handling serialization exception", e);
-            recordSendError(topic, e, null, context, processorNodeId);
-            return;
+            throw new FailedProcessingException("Fatal user code error in production error callback", fatalUserException);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:331]: fix: inside `send()` is seem incorrect to call `recordSendError()` which we only use in from the `producer.send(..., Calback)` which is called async. Here inside `RecordCollector.send()` (before we hit `producer.send()`) can should rather throw directly

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 10]
```diff
@@ -815,19 +815,25 @@ record = null;
...
-                record.offset(),
-                getStacktraceString(e)
+                record.offset()
             ),
-            e
-        );
+            originalException);
+    }
+
+    private void handleException(final String errorMessage, final Throwable originalException) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:837]: this is an attempt to preserve the error message we create when a handler throws by itself.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 11]
```diff
@@ -815,19 +815,25 @@ record = null;
         return true;
     }
 
-    private void handleException(final Throwable e) {
-        final StreamsException error = new StreamsException(
+    private void handleException(final Throwable originalException) {
+        handleException(
             String.format(
-                "Exception caught in process. taskId=%s, processor=%s, topic=%s, partition=%d, offset=%d, stacktrace=%s",
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:826]: given what we pass in `originalException` into `StreamsException, it seems not necessary to include the stracktrace in the error message.

**Comment by sebastienviale** [comment]
thanks for the PR, you're just ahead of me

**Comment by sebastienviale** [comment]
I just notice a nit in ProductionExceptionHandler 

     * @param record        the record that failed to serialize
     * @param exception     the exception that occurred during serialization
     * @deprecated Since 3.9. Use {@link #handle(ErrorHandlerContext, ProducerRecord, Exception)} instead.
     */
    @Deprecated
    default ProductionExceptionHandlerResponse handleSerializationException(final ProducerRecord record,
                                                                            final Exception exception) {
        return ProductionExceptionHandlerResponse.FAIL;
    }

should be

* @deprecated Since 3.9. Use {@link #handleSerializationException(ErrorHandlerContext, ProducerRecord, Exception, SerializationExceptionOrigin)} instead.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 11 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:826]: this seems to break two unit tests in ProcessingExceptionHandlerIntegrationTest while testing exception.getMessage

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:826]: Thanks! -- I did only run a few test locally, not the whole suite. Will take a look.

**Comment by loicgreffier** [comment]
A very little typo `Use Please` brought by KAFKA-16448 here: https://github.com/apache/kafka/blob/8d8c367066f59d888d6aa2011b840915cd07a3c2/streams/src/main/java/org/apache/kafka/streams/errors/DeserializationExceptionHandler.java#L40

**Comment by mjsax** [comment]
Updates this PR, and also extended it to align all 5 cases we have:
 - deserialzation
 - production / serialization
 - production / send
 - processing
 - punctuation
 
We now handle both error cases "handler throws exception" and "handler returns null" (this one I just realized while working on this PR and just added it).

File: streams/src/test/java/org/apache/kafka/streams/integration/ProcessingExceptionHandlerIntegrationTest.java [hunk 12]
```diff
@@ -222,9 +221,9 @@ public void shouldStopOnFailedProcessorWhenProcessingExceptionOccursInContinuePr
     }
 
     @Test
-    public void shouldStopProcessingWhenFatalUserExceptionInFailProcessingExceptionHandler() {
+    public void shouldStopProcessingWhenProcessingExceptionHandlerReturnsNull() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ProcessingExceptionHandlerIntegrationTest.java:224]: Re-purposed this test case for NULL case.

It seem redundant to test "throw exception" twice as originally in the code (one for FAIL and once for CONTINUE), because if we throw, there is no response we return at all...

File: streams/src/test/java/org/apache/kafka/streams/integration/ProcessingExceptionHandlerIntegrationTest.java [hunk 13]
```diff
@@ -308,9 +313,6 @@ public void configure(final Map<String, ?> configs) {
     public static class FailProcessingExceptionHandlerMockTest implements ProcessingExceptionHandler {
         @Override
         public ProcessingExceptionHandler.ProcessingHandlerResponse handle(final ErrorHandlerContext context, final Record<?, ?> record, final Exception exception) {
-            if (((String) record.key()).contains("FATAL")) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/ProcessingExceptionHandlerIntegrationTest.java:311]: Not needed any longer -- we just need one handler that can throw an exception

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java [hunk 14]
```diff
@@ -1242,7 +1244,7 @@ public void shouldThrowStreamsExceptionOnSubsequentSendIfFatalEvenWithContinueEx
             logContext,
             taskId,
             getExceptionalStreamsProducerOnSend(exception),
-            new ProductionExceptionHandlerMock(ProductionExceptionHandler.ProductionExceptionHandlerResponse.CONTINUE),
+            new ProductionExceptionHandlerMock(Optional.of(ProductionExceptionHandlerResponse.CONTINUE)),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1247]: Want to test 4 cases, so added `Optional` to make this possible.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordDeserializerTest.java [hunk 15]
```diff
@@ -62,27 +63,29 @@ public class RecordDeserializerTest {
...
-        final ConsumerRecord<Object, Object> record = recordDeserializer.deserialize(null, rawRecord);
-        assertEquals(rawRecord.topic(), record.topic());
-        assertEquals(rawRecord.partition(), record.partition());
-        assertEquals(rawRecord.offset(), record.offset());
-        assertEquals("key", record.key());
-        assertEquals("value", record.value());
-        assertEquals(rawRecord.timestamp(), record.timestamp());
-        assertEquals(TimestampType.CREATE_TIME, record.timestampType());
-        assertEquals(rawRecord.headers(), record.headers());
+        try (final Metrics metrics = new Metrics()) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordDeserializerTest.java:66]: Side cleanup. `Metrics` must be close to avoid resource leaks. (A few more times below.)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 16]
```diff
@@ -2661,17 +2662,20 @@ public void shouldNotCheckpointAfterRestorationWhenExactlyOnceEnabled() {
     }
 
     @Test
-    public void shouldPunctuateNotHandleFailProcessingExceptionAndThrowStreamsException() {
+    public void punctuateShouldNotHandleFailProcessingExceptionAndThrowStreamsException() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:2665]: renaming to make it readable :) (same below)

Plus some formatting changes to make the code readable

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 17]
```diff
@@ -2723,56 +2737,106 @@ public void shouldPunctuateNotHandleTaskMigratedExceptionAndThrowItAsIs() {
...
-        task = createStatelessTask(createConfig(AT_LEAST_ONCE, "100",
-            LogAndFailExceptionHandler.class.getName(), LogAndContinueProcessingExceptionHandler.class.getName()));
+        task = createStatelessTask(createConfig(
+            AT_LEAST_ONCE,
+            "100",
+            LogAndFailExceptionHandler.class.getName(),
+            LogAndContinueProcessingExceptionHandler.class.getName()
+        ));
 
-        assertDoesNotThrow(() ->
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:2732]: No need for this... if it would throw, the test fails anyway.

File: streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java:94]: I know it's late in the game to ask this, but what about a timestamp? Just a thought, I don't have a strong opinion here.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java [hunk 18]
```diff
@@ -1170,7 +1171,9 @@ public void process(final Record<Long, Long> record) {
         final KafkaStreams streams = new KafkaStreams(builder.build(), config);
 
         streams.setUncaughtExceptionHandler((t, e) -> {
-            if (uncaughtException != null || !e.getMessage().contains("Injected test exception")) {
+            if (uncaughtException != null ||
+                !(e instanceof StreamsException) ||
+                !e.getCause().getMessage().equals("Injected test exception.")) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosIntegrationTest.java:1176]: This change is required only, because the original test conditions was not "correct" anyway. The outer exception was already a `StreamsException` wrapping the `RuntimeException` the test throws. Testing the output exception for type `RuntimeException` does not make any sense anyway, as all non-checked exception are `RuntimeException` and the check passes...

The only actual change we get in this PR, is that the error message of the outer `StreamsException` changes -- originally, it was not set explicitly, and thus was set as `"RuntimeException: Injected test exception"` -- thus, even if we did check the "wrong/outer" error message the test passed. This PR now sets the `StreamsException` error message explicitly (what I don't consider a backward compatibility issue), but to this check, we need to check the error message of the wrapped exception (bonus, we can test for `equals` instead of `contains`).

File: streams/src/test/java/org/apache/kafka/streams/integration/EosV2UpgradeIntegrationTest.java [hunk 19]
```diff
@@ -947,8 +948,9 @@ public void process(final Record<Long, Long> record) {
             } else {
                 int exceptionCount = exceptionCounts.get(appDir);
                 // should only have our injected exception or commit exception, and 2 exceptions for each stream
-                if (++exceptionCount > 2 || !(e instanceof RuntimeException) ||
-                    !(e.getMessage().contains("test exception"))) {
+                if (++exceptionCount > 2 ||
+                    !(e instanceof StreamsException) ||
+                    !(e.getCause().getMessage().endsWith(" test exception."))) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosV2UpgradeIntegrationTest.java:953]: As above. However, we cannot use `equals` as the use different error messages for different situation when we throw.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 8 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:930]: Why a `TimeoutException` here? Could a `punctuate` potentially throw it?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java [hunk 20]
```diff
@@ -1800,7 +1900,10 @@ public ProductionExceptionHandlerResponse handle(final ErrorHandlerContext conte
                                                          final ProducerRecord<byte[], byte[]> record,
                                                          final Exception exception) {
             assertInputs(context, exception);
-            return response;
+            if (response == null) {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java]: I don't think the `response` can be null should this be `response.isEmpty`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordDeserializerTest.java [hunk 21]
```diff
@@ -212,7 +283,10 @@ public DeserializationHandlerResponse handle(final ErrorHandlerContext context,
             assertEquals(expectedRecord, record);
             assertInstanceOf(RuntimeException.class, exception);
             assertEquals("KABOOM!", exception.getMessage());
-            return response;
+            if (response == null) {
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordDeserializerTest.java:286]: same as above here

**Review by bbejeck — COMMENTED** [review]
Thanks for the PR @mjsax overall this looks good to me, I have a couple of minor comments.

File: streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java:94]: Good question. Sounds like a miss... \cc @sebastienviale @loicgreffier

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 8 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:930]: > Seems we forgot to handle this case entirely... we might call context.forward() inside of punctuate() and thus need to handle the same error cases as for node.process()...

We could end up in a sink node, and thus in `RecordCollectorImpl` / `producer.send()` which could throw a `TimeoutException`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java (hunk 20 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java]: `null` is on purpose. It's how we setup the code (it's not something the handler return).

In the end `response` is "what should the handler return in this test run", and I miss use it to say "handler should throw an exception". I could also add a new boolean flag `shouldThrowException` instead to not miss use `response` for this case.

Will update the PR.

File: streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java:94]: Side comment: I would not block this PR on this question, but do a follow up one, if we want to add ts.

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/errors/internals/DefaultErrorHandlerContext.java:94]: Yep I agree it shouldn't block, adding a timestamp in a follow on PR is fine.

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `3.9` branch.
